### DISCUSSION TITLE
Help troubleshooting... Too many cargo processes

### DISCUSSION THREAD

**Discussion by augustocdias** [mrcjkb/rustaceanvim#498] [Q&A] [answered]
So since a couple weeks I can barely use rust-analyzer in neovim. It got to a point that my machine went out of memory and restarted itself. After investigating I found out that whenever I change something, dozens of `cargo test` processes are automatically spawned... I have no idea what is causing this... Could it be rust-analyzer? How do I troubleshoot this and find the root cause? I have tried a minimal config with just rustacean but I couldn't reproduce... So the plugin is not likely the cause, but I don't know what else to look at. Could you give me some help on how to find the issue?

This is the command that is ran:

```
cargo test --no-run --package core-crypto --lib --message-format=json
```

**Comment by mrcjkb** [comment]
Hey :wave: 

> whenever I change something, dozens of cargo test processes are automatically spawned

This sounds to me like a rust-analyzer issue. If you can't reproduce it with a minimal config, I would suggest bisecting your rust-analyzer config.
You should be able to confirm it's caused by rust-analyzer by trying with the default config (i.e. passing in an empty table).

**Reply by augustocdias** [reply]
Yeah it happens with the default config. I wanted to find a way to reproduce it with a minimal config but I wasn't able so far... It just doesn't happen if I have only have rustacean and I have no idea which other plugin could be causing this

**Reply by mrcjkb** [reply]
In that case, I recommend bisecting your other plugins.
Do you use neotest?

**Reply by augustocdias** [reply]
I have it installed but I don't use it regularly

**Reply by augustocdias** [reply]
I've tried disabling it but nothing changed...

**Reply by mrcjkb** [reply]
hmm then the only thing left is to bisect the other plugins.

**Reply by augustocdias** [reply]
I've noticed the processes are spawned exactly after rustacean finishes checking the project...

**Reply by mrcjkb** [reply]
[possibly related](https://users.rust-lang.org/t/cargo-test-run-on-every-save/46204)?

**Reply by augustocdias** [reply]
> [possibly related](https://users.rust-lang.org/t/cargo-test-run-on-every-save/46204)?

Is there a way to disable flycheck with rustacean? I use clippy as build on rust analyser's settings

**Reply by mrcjkb** [reply]
rustaceanvim doesn't run flycheck automatically. That's all rust-analyzer behaviour/configuration.

**Reply by augustocdias** [reply]
Hi. I've just found out what is causing this behavior. When rustacean tries to configure dap and checks the server status it is spawning this bunch of processes. As soon as I set `dap.autoload_configurations` to false the issue disappeared. The reason I didn't find this in a minimal config is likely because I didn't set `codelldb` there, which would make rustacean ignore dap. If you want a codebase to reproduce: https://github.com/wireapp/core-crypto/tree/chore/lock This is a big refactor in progress and the tests are mostly not compiling yet (which could be the reason... not sure)

**Reply by mrcjkb** [reply]
I see.
rustaceanvim itself doesn't spawn any cargo processes.
It just issues a `rust-analyzer/runnables` LSP client request. It's rust-analyzer that spawns the processes.
I would suggest opening an issue with rust-analyzer if you don't want to set `dap.autoload_configurations` to `false`, as there's nothing that can be done on the rustaceanvim side.

**Reply by augustocdias** [reply]
That's not what I'm understanding from reading the code. You get the runnables result from the server and for each result you call start from the `dap` module. Inside you call `vim.system` which basically runs `cargo test --no-run`.

**Reply by augustocdias** [reply]
Is it supposed to be starting dap during lsp initialisation?

**Reply by mrcjkb** [reply]
Oh yes, I forgot about that.
The issue is that if you have a lot of test configurations, it won't block on the `vim.system` call, leading to lots of parallel executions.

**Reply by mrcjkb** [reply, marked as answer]
This should solve the issue of parallel builds:

- https://github.com/mrcjkb/rustaceanvim/pull/501

> Is it supposed to be starting dap during lsp initialisation?

It doesn't actually run tests, it just compiles (hence the `--no-run` flag) and analyses the json output (hence the `--message-format=json` flag) to create a DAP configuration that can be registered with nvim-dap, so you can use commands like `DapContinue` to debug runnables discovered by rust-analyzer.

**Reply by augustocdias** [reply]
> Oh yes, I forgot about that.
> 
> The issue is that if you have a lot of test configurations, it won't block on the `vim.system` call, leading to lots of parallel executions.

It is not supposed to block but because of the amount it does generate a couple freezes. This also adds up as it is ran every time I change something and save.

**Reply by mrcjkb** [reply]
The latest release runs the builds sequentially (see the PR linked in my previous comment).

> This also adds up as it is ran every time I change something and save.

That doesn't sound right. It only runs when a new rust-analyzer client attaches.

**Reply by augustocdias** [reply]
> That doesn't sound right. It only runs when a new rust-analyzer client attaches.

Yes. But imagine hundreds of child processes running in the background. And not cheap ones, rust compiling takes quite some time

which leads to the question: does it need to run for every runnable returned by the analyzer? Wouldn't a single cargo test return all the metadata needed here? I don't see any parameter related to each runnable being passed so isn't the output of all this processes the same?

your pr did solve the problem btw, but I didn't see the cargo test being ran at all and I get no runnables or debuggables

**Reply by mrcjkb** [reply]
> And not cheap ones, rust compiling takes quite some time

cargo reuses artefacts that have been built already.
So it shouldn't be expensive with sequential builds.

> does it need to run for every runnable returned by the analyzer?

Yes. rustaceanvim has no idea what build commands it's getting from rust-analyzer.

> but I didn't see the cargo test being ran at all

There's no test being run (see the `--no-run` flag).

> and I get no runnables or debuggables

You mentioned earlier that your tests weren't compiling yet. Perhaps that's the reason. 
When building configurations to automatically add to nvim-dap, it fails silently.

**Reply by augustocdias** [reply]
>cargo reuses artefacts that have been built already.
So it shouldn't be expensive with sequential builds.

I think that's the reason it was starving my machine before. Imagine cargo trying to build in parallel hundreds of times and one waiting for the next to lock the build on the project. That definitely has to run sequentially.

> There's no test being run (see the --no-run flag).

I meant `cargo test --no-run` itself.

> You mentioned earlier that your tests weren't compiling yet. Perhaps that's the reason.
When building configurations to automatically add to nvim-dap, it fails silently.

Yeah... I'll try later with a project that is working.
